File: 算法/排序/希尔排序.py
def shell_sort(unordered_ls: list) -> None:
    n = len(unordered_ls)
    gap = n // 2
    while gap > 0:  # gap == 1 的时候，相当于直接插入排序
        for i in range(gap, n):
            key = unordered_ls[i]
            j = i - gap
            while (j >= 0 and unordered_ls[j] > key):
                unordered_ls[j + gap] = unordered_ls[j]
                j -= gap            
            unordered_ls[j + gap] = key
        gap //= 2

if __name__ == "__main__":
    ls = [30,120, 89, 1700, 78, 1290, 6790, 1888, 92, 109, 10, 40, 20, 50]
    shell_sort(ls)
    for i in ls:
        print(i, end=" ")

File: 算法/排序/希尔排序.cpp
//
// Created by Momo on 2023/9/14.
//
#include <iostream>
#include <vector>

using namespace std;

template<typename T>
static void shell_sort(vector<T> &v) {
    // 每一次的跳跃量
    for (int gap = v.size() / 2; gap > 0; gap /= 2) {
        // 把跳跃间隔之间的变量作为一个 "简单插入排序" 的子列
        for (int i = gap; i < v.size(); i++) {
            T key = v.at(i);
            int j = i - gap;    // 寻找 key 前面的元素
            while (j >= 0 and v.at(j) > key) {
                v.at(j + gap) = std::move(v.at(j)); // 向右移动
                j -= gap;
            }
            v.at(j + gap) = key;
        }
    }
}

int main() {
    vector<int> arr1 = {12, 11, 13, 5, 6};
    shell_sort(arr1);
    for (const auto &num: arr1)
        cout << num << " ";
    cout << endl;
    return 0;
}